lib/command/src/Obelisk/Command/Upgrade.hs

```diff
@@ -0,0 +1,186 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Obelisk.Command.Upgrade where
+
+import Control.Monad (forM_, unless, void)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Trans.Maybe (MaybeT (..), runMaybeT)
+import qualified Data.Map as Map
+import Data.Semigroup ((<>))
+import Data.Text (Text)
+import qualified Data.Text as T
+import System.Environment (getExecutablePath)
+import System.FilePath
+import System.Posix.Process (executeFile)
+import System.Process (proc)
+
+import Obelisk.App (MonadObelisk)
+import Obelisk.CliApp
+import Obelisk.Command.Utils
+
+import Obelisk.Command.Project (toImplDir)
+import Obelisk.Command.Project (findProjectObeliskCommand)
+import Obelisk.Command.Thunk (updateThunk)
+
+import Obelisk.Migration
+
+data MigrationGraph
+  = MigrationGraph_ObeliskUpgrade
+  | MigrationGraph_ObeliskHandoff
+  deriving (Eq, Show)
+
+graphName :: MigrationGraph -> Text
+graphName = \case
+  MigrationGraph_ObeliskHandoff -> "obelisk-handoff"
+  MigrationGraph_ObeliskUpgrade -> "obelisk-upgrade"
+
+fromGraphName :: Text -> MigrationGraph
+fromGraphName = \case
+  "obelisk-handoff" -> MigrationGraph_ObeliskHandoff
+  "obelisk-upgrade" -> MigrationGraph_ObeliskUpgrade
+  _ -> error "Invalid graph name specified"
+
+
+ensureCleanProject :: MonadObelisk m => FilePath -> m ()
+ensureCleanProject project = ensureCleanGitRepo project False "Cannot upgrade with uncommited changes"
```

Comment:
Is there a specific reason we can't upgrade when there are uncommitted changes?  And, if so, why do we not care about ignored changes?

Reply:
The only files `ob upgrade` will touch in the project repo is the `.obelisk/impl` thunk. 

At minimum we want that path to be clean, but other than that there is no reason to require the rest of the repository to be clean. I just figured it is easier to reuse that function instead of writing another that checks just for the subpath. Should we do the specific check?